File: Leetcode/Python/_5.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        dp = [[False] * n for _ in range(n)]
        max_len = 1
        output = '' if not s else s[0]
        for i in range(n - 1, -1, -1):
            dp[i][i] = True
            for j in range(i + 1, n):
                if s[i] == s[j]:
                    if j - i == 1 or dp[i + 1][j - 1]:
                        dp[i][j] = True
                        if max_len <= j - i + 1:
                            max_len = j - i + 1
                            output = s[i:j + 1]
        return output


class Solution:
    def longestPalindrome(self, s):
            longest_palindrom = ''
            dp = [[0]*len(s) for _ in range(len(s))]
            for i in range(len(s)):
                dp[i][i] = 1
                longest_palindrom = s[i]

            for i in range(len(s)-1, -1, -1):
                for j in range(i+1,len(s)):  
                    if s[i] == s[j]:
                        if j-i ==1 or dp[i+1][j-1]:
                            dp[i][j] = 1
                            if len(longest_palindrom) < len(s[i:j+1]):
                                longest_palindrom = s[i:j+1]
            return longest_palindrom

File: Leetcode/cpp/5.cpp
#include <iostream>
#include <vector>
#include <cassert>
#include <string>
#include <climits>

class Solution {
public:
  std::string longestPalindrome(std::string s) {
    int n = static_cast<int>(s.size());

    if (n == 0) return "";

    std::vector<std::vector<int>> dp(n, std::vector<int>(n, 0));

    int max_len = 1;
    int start = 0;
    for (int i = n - 1; i >= 0; --i) {
      dp[i][i] = 1;
      for (int j = i + 1; j < n; ++j) {
        if (s[i] == s[j] && (j - i == 1 || dp[i + 1][j - 1])) {
          dp[i][j] = 1;
          int len = j + 1 - i;
          if (max_len <= len) {
            max_len = len;
            start = i;
          }
        }
      }
    }
    // for (int i = 0; i < n; ++i) {
    //   for (int j = 0; j < n; ++j) {
    //     std::cout << dp[i][j] << " ";
    //   }
    //   std::cout << std::endl;
    // }
    return s.substr(start, max_len); 
  }
};

int main() {
  Solution sol;
  std::string s = "babad";
  std::vector<std::string> expected_output = { "aba", "bab"};
  std::string output = sol.longestPalindrome(s);
  bool found = false;
  int i = 0;
  for ( ; i < expected_output.size(); ++i) {
    if (output == expected_output[i]) {
      found = true;
      break;
    }
  }
  if (!found) {
    std::cout << "Not found solution!" << std::endl;
  }
  std::cout << "Tests are passed!" << std::endl;
  
  return 0;
}
